qiskit/utils/deprecation.py

```diff
@@ -77,3 +92,73 @@ def _rename_kwargs(func_name, kwargs, kwarg_map, category: Type[Warning] = Depre
                 )
 
                 kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _extend_docstring(func, version, kwarg_map):
+    """kwarg_map[None] means message in no kwarg, and it will be
+    append to the long desscription"""
+    docstr = func.__doc__
+    if docstr:
+        docstr_lines = docstr.expandtabs().splitlines()
+    else:
+        docstr_lines = ["DEPRECATED"]
+
+    indent = 1000
+    first_empty_line = None
+    for line_no, line in enumerate(docstr_lines[1:], start=1):
+        stripped = line.lstrip()
+        if stripped:
+            indent = min(indent, len(line) - len(stripped))
+        else:
+            if first_empty_line is None:
+                first_empty_line = line_no
+    if first_empty_line is None:
+        first_empty_line = len(docstr_lines)
+    spaces = ""
+    if indent != 1000:
+        spaces = " " * indent
+
+    new_doc_str_lines = docstr_lines[:first_empty_line]
+    if None in kwarg_map:
+        new_doc_str_lines += ["", spaces + f".. deprecated:: {version}"]
+        for msg_line in kwarg_map[None]:
+            new_doc_str_lines.append(spaces + "  " + msg_line)
+
+    arg_indent = indent
+    args_section = False
+    deprecated_arg = False
+    for docstr_line in docstr_lines[first_empty_line:]:
+        stripped = docstr_line.lstrip()
+        current_indent = len(docstr_line) - len(stripped)
+        if args_section:
+            if deprecated_arg and current_indent == arg_indent:
+                new_doc_str_lines.append(docstr_line)
+                deprecated_arg = False
+            else:
+                if not deprecated_arg:
+                    for k in kwarg_map.keys():
+                        if k is None:
+                            continue
+                        if stripped.startswith(k):
```

Comment:
Rather than iterating through the whole dictionary each time (although admittedly it should be _too_ expensive), you might be better parsing the argument name from the string.  Extracting the text up to the first non-identifier character should be sufficient here.  You could have a compiled regex in the module scope `identifier = re.compile('[a-zA-Z_][a-zA-Z0-9_]*')`, then `identifier.match(line[arg_indent:])` would return the argument name if it matches (`re.match` requires the string to start with the match, so it will handle indented continuation lines correctly).

If you do that, you can just do a dict lookup, rather than iterating.